kupyna/src/long_compress.rs

```diff
@@ -1,95 +1,75 @@
-use crate::utils::{add_constant_plus, add_constant_xor, apply_s_box, mix_columns, xor_bytes};
+use crate::utils::{add_constant_plus, add_constant_xor, apply_s_box, mix_columns, xor_words};
 
 pub(crate) const COLS: usize = 16;
 const ROUNDS: u64 = 14;
 
-type Matrix = [[u8; 8]; 16];
-
 pub(crate) fn compress(prev_vector: &mut [u64; COLS], message_block: &[u8; 128]) {
-    let mut prev_vector_u8 = [0u8; 128];
-    for (src, dst) in prev_vector.iter().zip(prev_vector_u8.chunks_exact_mut(8)) {
-        dst.copy_from_slice(&src.to_be_bytes());
+    // Convert message block from u8 to u64 (column-major order as per paper)
+    let mut message_u64 = [0u64; COLS];
+    for (chunk, v) in message_block.chunks_exact(8).zip(message_u64.iter_mut()) {
+        *v = u64::from_be_bytes(chunk.try_into().unwrap());
     }
 
-    let m_xor_p = xor_bytes(*message_block, prev_vector_u8);
+    let m_xor_p = xor_words(*prev_vector, message_u64);
 
     let t_xor_mp = t_xor_l(m_xor_p);
 
-    let t_plus_m = t_plus_l(*message_block);
-
-    prev_vector_u8 = xor_bytes(xor_bytes(t_xor_mp, t_plus_m), prev_vector_u8);
+    let t_plus_m = t_plus_l(message_u64);
 
-    for (dst, src) in prev_vector.iter_mut().zip(prev_vector_u8.chunks_exact(8)) {
-        *dst = u64::from_be_bytes(src.try_into().unwrap());
-    }
+    *prev_vector = xor_words(xor_words(t_xor_mp, t_plus_m), *prev_vector);
 }
 
-pub(crate) fn t_plus_l(block: [u8; 128]) -> [u8; 128] {
-    let mut state = block_to_matrix(block);
+pub(crate) fn t_plus_l(state: [u64; COLS]) -> [u64; COLS] {
+    let mut state = state;
     for nu in 0..ROUNDS {
         state = add_constant_plus(state, nu as usize);
         state = apply_s_box(state);
         state = rotate_rows(state);
         state = mix_columns(state);
     }
-    matrix_to_block(state)
+    state
```

Comment:
It may be worth to change `add_constant_plus`, `apply_s_box`, `rotate_rows`, and `mix_columns` to work in-place, i.e. over `&mut [u64; COLS]`.